### PR TITLE
[protobuf] Add haskell binding and initial client library

### PR DESCRIPTION
**Pull request by TristanCacqueray**
This change adds a new monocle haskell project to enable
lentille integration using protobuf.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -4,15 +4,20 @@
...
 	mkdir -p srcgen/
-	(cd codegen; cabal run monocle-codegen ../protos/monocle/http.proto ../monocle/webapi.py ../srcgen/WebApi.res)
+	(cd codegen; cabal run monocle-codegen ../protos/monocle/http.proto ../src/Monocle/WebApi.hs ../monocle/webapi.py ../srcgen/WebApi.res)
+	~/.cabal/bin/ormolu -i ./src/Monocle/WebApi.hs
 	black ./monocle/webapi.py
 	./web/node_modules/.bin/bsc -format ./srcgen/WebApi.res > ./web/src/components/WebApi.res
 	rm -Rf srcgen/
 
+codegen-haskell:
+	sh -c 'for pb in $(MESSAGES); do ~/.cabal/bin/compile-proto-file --includeDir /usr/include --includeDir protos/ --proto $${pb} --out src/; done'
```
Comment (morucci) [inline Makefile:18]: could we output in haskell/src instead ?

File: codegen/rename_bs_module.py [hunk 2]
```diff
@@ -34,7 +34,13 @@ def fix_module(filepath):
         content = content.replace("Regex", "_regex")
         print("Replaced", typeName, newTypeName)
         content = content.replace(typeName, newTypeName)
-    newFile.write_text(content)
+    newFile.write_text(
+        content.replace(
+            # Fix timestamp message encoding which is a rfc3339 string, not an object
+            'timestamp" (Js.Json.object_',
+            'timestamp" (Js.Json.string',
```
Comment (morucci) [inline codegen/rename_bs_module.py:41]: So we keep the protoc python encoding for timestamp and adapt the rest ?

File: doc/openapi.yaml [hunk 3]
```diff
@@ -24,6 +24,23 @@ paths:
                         application/json:
                             schema:
                                 $ref: '#/components/schemas/GetProjectsResponse'
+    /api/1/task_data_add:
```
Comment (morucci) [inline doc/openapi.yaml:27]: Why not having something like ?:
- /api/1/task_data/add
- /api/1/task_data/commit
- /api/1/task_data/get_last_updated]
- /api/1/projects/get

File: src/Monocle/Worker.hs [hunk 4]
```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE DerivingStrategies #-}
```
Comment (morucci) [inline src/Monocle/Worker.hs:1]: I see this more tied to lentille ...

File: codegen/rename_bs_module.py (hunk 2 above)
Comment (TristanCacqueray) [reply codegen/rename_bs_module.py:41]: Yes, the code generated by the protobuf default compiler does a special treatment for timestamp, so we only have to adapt rescript and haskell encoders.

File: Makefile (hunk 1 above)
Comment (TristanCacqueray) [reply Makefile:18]: Sure.

File: .github/workflows/haskell.yaml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+name: Haskell
```
Comment (TristanCacqueray) [inline .github/workflows/haskell.yaml:1]: TODO: check that workflow works

File: haskell/src/Google/Protobuf/Timestamp.hs [hunk 6]
```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE DataKinds #-}
```
Comment (TristanCacqueray) [inline haskell/src/Google/Protobuf/Timestamp.hs:1]: TODO: replace by https://github.com/awakesecurity/proto3-suite/pull/150 when available on hackage.

File: haskell/src/Monocle/Worker.hs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+  startTime <- log' LogStarting
+  since <- case sinceM of
+    Just ts -> pure ts
+    Nothing -> getTimestampFromApi
+  process (retry . taskDataAdd monocleClient . mkRequest) (runFetcher tdf since)
+  res <- retry $ commitTimestamp startTime
+  log (if res then LogEnded else LogFailed)
+  where
+    commitTimestamp startTime = do
+      -- setUpdatedSince monocleClient indexName crawlerName apiKey startTime
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Worker.hs:166]: nit: remove lentille left-over

File: protos/monocle/task_data.proto [hunk 8]
```diff
@@ -44,3 +45,30 @@ message TaskDataGetLastUpdatedResponse {
...
+  string crawler = 2;
+  string apikey = 3;
+  repeated NewTaskData items = 4;
+}
+
+message AddResponse {
+  oneof result {
+    // TODO: define a header message with generic error
+    // then do not re-use the commit error enum.
+    TaskDataCommitError error = 1;
```
Comment (TristanCacqueray) [inline protos/monocle/task_data.proto:72]: TODO: use a better error message, e.g. using a common header field, and more specific add errors such as ToManyTaskData.

**Review by TristanCacqueray — COMMENTED** [review]
LGTM

**Review by morucci — APPROVED** [review]

**Pull request merged**